cf_remote/utils.py

```diff
@@ -226,15 +228,64 @@ def print_progress_dot(*args):
     sys.stdout.flush()  # STDOUT is line-buffered
 
 
+# atomic copy, see lock-free whack-a-mole algorithm
 def copy_file(input_path, output_path):
-    filename = os.path.basename(input_path)
-    output_dir = os.path.dirname(output_path)
-
-    tmp_filename = ".{}.tmp".format(filename)
-    tmp_output_path = os.path.join(output_dir, tmp_filename)
-
-    shutil.copyfile(input_path, tmp_output_path)
-    os.rename(tmp_output_path, output_path)
+    if os.path.exists(output_path):
+        return
+
+    output_filename = os.path.basename(output_path)
+    if not output_filename:
+        output_filename = os.path.basename(input_path)
+
+    output_dirname = os.path.dirname(output_path)
+    my_uuid = uuid.uuid4()
+
+    tmp_path = os.path.join(
+        output_dirname, "{}-{}.tmp".format(output_filename, my_uuid)
+    )
+    shutil.copyfile(input_path, tmp_path)
+
+    my_mole_path = os.path.join(
+        output_dirname, "{}-{}.mole.tmp".format(output_filename, my_uuid)
+    )
+    os.rename(tmp_path, my_mole_path)
+
+    glob_pattern = "{}-*.mole.tmp".format(output_filename)
+    re_pattern = r"{}-(.*)\.mole\.tmp".format(output_filename)
+
+    def whack_a_mole(my_uuid, my_mole_path):
+        moles = glob.glob(os.path.join(output_dirname, glob_pattern))
+        for mole in moles:
+            match = re.search(re_pattern, mole)
+            assert match is not None
+            mole_uuid = uuid.UUID(match.group(1))
+
+            if mole_uuid > my_uuid:
```

Comment:
Since the part of the file before and after the UID are the same, you may as well compare the whole string. E.g.
```python3
>>> foo = "foo.001.mole"
>>> bar = "foo.002.mole"
>>> foo > bar
False
>>> bar > foo
True
```
Just make sure that it starts with the base name and ends with `.mole`